addons/mail/static/src/discuss/call/common/call_context_menu.js

```diff
@@ -6,51 +6,132 @@ import { browser } from "@web/core/browser/browser";
 
 export class CallContextMenu extends Component {
     static props = ["rtcSession", "close?"];
     static template = "discuss.CallContextMenu";
 
     updateStatsTimeout;
+    rtcConnectionTypes = CONNECTION_TYPES;
 
     setup() {
         this.userSettings = useState(useService("mail.user_settings"));
+        this.rtc = useState(useService("discuss.rtc"));
+        this.state = useState({
+            downloadStats: {},
+            uploadStats: {},
+            producerStats: {},
+        });
         onMounted(() => {
             if (!this.env.debug) {
                 return;
             }
-            this.props.rtcSession.updateStats();
-            this.updateStatsTimeout = browser.setInterval(
-                () => this.props.rtcSession.updateStats(),
-                3000
-            );
+            this.updateStats();
+            this.updateStatsTimeout = browser.setInterval(() => this.updateStats(), 3000);
         });
         onWillUnmount(() => browser.clearInterval(this.updateStatsTimeout));
     }
 
+    get isSelf() {
+        return this.rtc.state.selfSession?.eq(this.props.rtcSession);
+    }
+
     get inboundConnectionTypeText() {
-        if (!this.props.rtcSession.remoteCandidateType) {
-            return _t("no connection");
-        }
-        return _t("%(candidateType)s (%(protocol)s)", {
-            candidateType: this.props.rtcSession.remoteCandidateType,
-            protocol: PROTOCOLS_TEXT[this.props.rtcSession.remoteCandidateType],
-        });
+        const candidateType =
+            this.rtc.state.connectionType === CONNECTION_TYPES.SERVER
+                ? this.state.downloadStats.remoteCandidateType
+                : this.props.rtcSession.remoteCandidateType;
+        return this.formatProtocol(candidateType);
     }
 
     get outboundConnectionTypeText() {
-        if (!this.props.rtcSession.localCandidateType) {
+        const candidateType =
+            this.rtc.state.connectionType === CONNECTION_TYPES.SERVER
+                ? this.state.uploadStats.localCandidateType
+                : this.props.rtcSession.localCandidateType;
+        return this.formatProtocol(candidateType);
+    }
+
+    get volume() {
+        return this.userSettings.getVolume(this.props.rtcSession);
+    }
+
+    /**
+     * @param {string} candidateType
+     * @returns {string} a formatted string that describes the connection type e.g: "prflx (STUN)"
+     */
+    formatProtocol(candidateType) {
+        if (!candidateType) {
             return _t("no connection");
         }
         return _t("%(candidateType)s (%(protocol)s)", {
-            candidateType: this.props.rtcSession.localCandidateType,
-            protocol: PROTOCOLS_TEXT[this.props.rtcSession.localCandidateType],
+            candidateType,
+            protocol: PROTOCOLS_TEXT[candidateType],
         });
     }
 
-    get volume() {
-        return this.userSettings.getVolume(this.props.rtcSession);
+    async updateStats() {
```

Comment:
I am not sure how often will this function be called. I think the assigning part can be handled separately since it is pretty costly. ( if within if and switch in the loop, etc.) So that the assigning part will not block the following working flow.

Reply:
And also, since it is costly. I think it is a good idea to keep a structure that only updates the last request and drop the previous updates from the same person if there is a lot of unfinished rtc. I am not sure about the structure, something like `keepLast` I guess?

Reply:
there is no RPC done here, `sfuClient.getStats()` will obtain the stats from the upload and download medisaoup transports, which are obtained from the [`getStats()` of their respective underlying local RTCPeerConnection](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/getStats)

Reply:
The only way to call it often is to quickly open and close the call context menu, but if you do that, you generate a new component (the whole menu), so only the latest component will exist and re-render when the latest promise is resolved.